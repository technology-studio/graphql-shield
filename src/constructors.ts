import * as Yup from 'yup'
import type { ValidateOptions } from 'yup'
import {
  IRuleFunction,
  IRuleConstructorOptions,
  ShieldRule,
  IShieldContext,
} from './types'
import {
  Rule,
  RuleAnd,
  RuleOr,
  RuleNot,
  RuleTrue,
  RuleFalse,
  InputRule,
  RuleChain,
  RuleRace,
} from './rules'

/**
 *
 * @param name
 * @param options
 *
 * Wraps a function into a Rule class. This way we can identify rules
 * once we start generating middleware from our ruleTree.
 *
 * 1.
 * const auth = rule()(async (parent, args, ctx, info) => {
 *  return true
 * })
 *
 * 2.
 * const auth = rule('name')(async (parent, args, ctx, info) => {
 *  return true
 * })
 *
 * 3.
 * const auth = rule({
 *  name: 'name',
 *  fragment: 'string',
 *  cache: 'cache',
 * })(async (parent, args, ctx, info) => {
 *  return true
 * })
 *
 */
export const rule = (
  name?: string | IRuleConstructorOptions,
  options?: IRuleConstructorOptions,
) => (func: IRuleFunction): Rule => {
  if (typeof name === 'object') {
    options = name
    name = Math.random().toString()
  } else if (typeof name === 'string') {
    options = options || {}
  } else {
    name = Math.random().toString()
    options = {}
  }

  return new Rule(name, func, {
    fragment: options.fragment,
    cache: options.cache,
  })
}

/**
 *
 * Constructs a new InputRule based on the schema.
 *
 * @param schema
 */
export const inputRule = <T>(name?: string) => (
  schema: (yup: typeof Yup, ctx: IShieldContext) => Yup.Schema<T>,
  options?: ValidateOptions,
) => {
  if (typeof name === 'string') {
    return new InputRule(name, schema, options)
  } else {
    return new InputRule(Math.random().toString(), schema, options)
  }
}

/**
 *
 * @param rules
 *
 * Logical operator and serves as a wrapper for and operation.
 *
 */
export const and = (...rules: ShieldRule[]): RuleAnd => {
  return new RuleAnd(rules)
}

/**
 *
 * @param rules
 *
 * Logical operator and serves as a wrapper for and operation.
 *
 */
export const chain = (...rules: ShieldRule[]): RuleChain => {
  return new RuleChain(rules)
}

/**
 *
 * @param rules
 *
 * Logical operator and serves as a wrapper for and operation.
 *
 */
export const race = (...rules: ShieldRule[]): RuleRace => {
  return new RuleRace(rules)
}

/**
 *
 * @param rules
 *
 * Logical operator or serves as a wrapper for or operation.
 *
 */
export const or = (...rules: ShieldRule[]): RuleOr => {
  return new RuleOr(rules)
}

/**
 *
 * @param rule
 *
 * Logical operator not serves as a wrapper for not operation.
 *
 */
export const not = (rule: ShieldRule, error?: string | Error): RuleNot => {
  if (typeof error === 'string') return new RuleNot(rule, new Error(error))
  return new RuleNot(rule, error)
}

/**
 *
 * Allow queries.
 *
 */
export const allow = new RuleTrue()

/**
 *
 * Deny queries.
 *
 */
export const deny = new RuleFalse()
