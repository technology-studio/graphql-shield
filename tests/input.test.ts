import { graphql } from 'graphql'
import { applyMiddleware } from '@txo/graphql-middleware'
import { makeExecutableSchema } from '@graphql-tools/schema'
import { shield, inputRule } from '../src'

describe('input rule', () => {
  test('schema validation works as expected', async () => {
    const typeDefs = `
      type Query {
        hello: String!
      }

      type Mutation {
        login(email: String!): String
      }
    `

    const resolvers = {
      Query: {
        hello: () => 'world',
      },
      Mutation: {
        login: () => 'pass',
      },
    }

    const schema = makeExecutableSchema({ typeDefs, resolvers })

    // Permissions
    const permissions = shield({
      Mutation: {
        login: inputRule()((yup) =>
          yup.object({
            email: yup.string().email('It has to be an email!').required(),
          }),
        ),
      },
    })

    const schemaWithPermissions = applyMiddleware(schema, permissions)

    /* Execution */

    const query = `
      mutation {
        success: login(email: "shield@graphql.com")
        failure: login(email: "notemail")
      }
    `
    const res = await graphql({
      schema: schemaWithPermissions, 
      source: query, 
    })

    /* Tests */

    expect(res.data).toEqual({
      success: 'pass',
      failure: null,
    })
    expect(res.errors).toMatchSnapshot()
  })
})
