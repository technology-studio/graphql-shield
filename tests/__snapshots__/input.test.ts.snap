// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`input rule schema validation works as expected 1`] = `
[
  [GraphQLError: It has to be an email!],
]
`;
